core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -489,56 +491,94 @@ protected int comparePivot(int j) {
         }
 
         /**
-         * This function executes a move operation moving the given shard from
-         * the given node to the minimal eligible node with respect to the
-         * weight function. Iff the shard is moved the shard will be set to
+         * Move started shards that can not be allocated to a node anymore
+         *
+         * For each shard to be moved this function executes a move operation
+         * to the minimal eligible node with respect to the
+         * weight function. If a shard is moved the shard will be set to
          * {@link ShardRoutingState#RELOCATING} and a shadow instance of this
          * shard is created with an incremented version in the state
          * {@link ShardRoutingState#INITIALIZING}.
          *
-         * @return <code>true</code> iff the shard has successfully been moved.
+         * @return <code>true</code> if the allocation has changed, otherwise <code>false</code>
          */
-        public boolean move(ShardRouting shard, RoutingNode node ) {
-            if (nodes.isEmpty() || !shard.started()) {
-                /* with no nodes or a not started shard this is pointless */
+        public boolean moveShards() {
+            if (nodes.isEmpty()) {
+                /* with no nodes this is pointless */
                 return false;
             }
-            if (logger.isTraceEnabled()) {
-                logger.trace("Try moving shard [{}] from [{}]", shard, node);
+
+            // create a copy of the shards interleaving between nodes, and check if they can remain
+            List<ShardRouting> shards = new ArrayList<>();
+            int index = 0;
+            boolean found = true;
+            final RoutingNodes routingNodes = allocation.routingNodes();
+            while (found) {
```

Comment:
The iteration order is a very interesting one currently: We take one shard from each node in succession (and start again on the first node as long as there are still shards). I don't know exactly why, but I guess it is to make it fairer so that every node gets a chance to get rid of some shards (i.e. a more "balanced" riddance of shards)? I'm not sure whether we really need this and if it has any effect at all. It is not even truly fair as it creates the order independently on whether a shard is later moved away. 

I was a bit hesitant on changing the semantics in this PR though.

@kimchy Can you clarify why this interleaving between nodes was done and whether you believe it to be still useful?